File: src/test/java/de/gematik/test/fhir/TestResourceSupport.java
package de.gematik.test.fhir;

import lombok.val;
import org.hl7.fhir.r4.formats.JsonParser;
import org.hl7.fhir.r4.formats.XmlParser;
import org.hl7.fhir.r4.model.Resource;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Locale;

final class TestResourceSupport {

  private static final XmlParser XML_PARSER = new XmlParser();
  private static final JsonParser JSON_PARSER = new JsonParser();

  private TestResourceSupport() {
  }

  static Resource readXmlResource(Path path) throws IOException {
    return XML_PARSER.parse(Files.readString(path, StandardCharsets.UTF_8));
  }

  static Resource readJsonResource(Path path) throws IOException {
    return JSON_PARSER.parse(Files.readString(path, StandardCharsets.UTF_8));
  }

  static List<Resource> loadFourSourceResources(Path baseDir) throws IOException {
    val sources = new ArrayList<Resource>();
    try (val paths = Files.walk(baseDir)) {
      paths
          .filter(Files::isRegularFile)
          .filter(TestResourceSupport::isXmlOrJsonFile)
          .sorted(Comparator.comparing(Path::toString))
          .forEach(path -> sources.add(readResource(path)));
    }
    return sources;
  }

  private static boolean isXmlOrJsonFile(Path path) {
    val fileName = path.getFileName().toString().toLowerCase(Locale.ROOT);
    return fileName.endsWith(".xml") || fileName.endsWith(".json");
  }

  private static Resource readResource(Path path) {
    try {
      if (path.getFileName().toString().toLowerCase(Locale.ROOT).endsWith(".xml")) {
        return readXmlResource(path);
      }
      return readJsonResource(path);
    } catch (IOException e) {
      throw new IllegalStateException("Unable to read test resource: " + path, e);
    }
  }


  static void maybeWriteResource(Resource resource, Path outputPath) throws IOException {
    if (!Boolean.getBoolean("writeArtifacts")) {
      return;
    }
    Files.createDirectories(outputPath.getParent());
    Files.writeString(outputPath, JSON_PARSER.composeString(resource), StandardCharsets.UTF_8);
  }
}

File: src/test/kotlin/de/gematik/test/fhir/TestResourceSupport.kt
package de.gematik.test.fhir

import org.hl7.fhir.r4.formats.JsonParser
import org.hl7.fhir.r4.formats.XmlParser
import org.hl7.fhir.r4.model.Resource
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path
import java.util.Locale
import kotlin.io.path.extension
import kotlin.io.path.notExists

internal object TestResourceSupport {
    private const val WRITE_ARTIFACTS_PROPERTY = "writeArtifacts"
    private const val WRITE_ARTIFACTS_ENV = "WRITE_ARTIFACTS"
    private const val SUREFIRE_MARKER_PROPERTY = "surefire.real.class.path"

    private val xmlParser = XmlParser()
    private val jsonParser = JsonParser()
    private val projectDir: Path = System.getProperty("basedir")
        ?.takeIf(String::isNotBlank)
        ?.let(Path::of)
        ?: Path.of(System.getProperty("user.dir"))
    private val buildDir: Path = System.getProperty("project.build.directory")
        ?.takeIf(String::isNotBlank)
        ?.let(Path::of)
        ?: projectDir.resolve("target")

    @Throws(IOException::class)
    fun readXmlResource(path: Path): Resource =
        xmlParser.parse(Files.readString(path, StandardCharsets.UTF_8))

    @Throws(IOException::class)
    fun readJsonResource(path: Path): Resource =
        jsonParser.parse(Files.readString(path, StandardCharsets.UTF_8))

    @Throws(IOException::class)
    fun loadFourSourceResources(baseDir: Path): List<Resource> =
        Files.walk(baseDir).use { paths ->
            paths
                .filter(Files::isRegularFile)
                .filter(::isXmlOrJsonFile)
                .sorted(compareBy(Path::toString))
                .map(::readResource)
                .toList()
        }

    private fun isXmlOrJsonFile(path: Path): Boolean =
        path.extension.lowercase(Locale.ROOT) in setOf("xml", "json")

    private fun readResource(path: Path): Resource =
        try {
            if (path.extension.lowercase(Locale.ROOT) == "xml") {
                readXmlResource(path)
            } else {
                readJsonResource(path)
            }
        } catch (exception: IOException) {
            throw IllegalStateException("Unable to read test resource: $path", exception)
        }

    fun buildArtifactPath(fileName: String): Path = buildDir.resolve(fileName)

    @Throws(IOException::class)
    fun writeResource(resource: Resource, outputPath: Path) {
        if (!isWriteArtifactsEnabled()) return
        outputPath.parent
            ?.takeIf(Path::notExists)
            ?.let(Files::createDirectories)
        Files.writeString(outputPath, jsonParser.composeString(resource), StandardCharsets.UTF_8)
    }

    private fun isWriteArtifactsEnabled(): Boolean =
        System.getProperty(WRITE_ARTIFACTS_PROPERTY)?.trim()?.equals("true", ignoreCase = true) ?: false ||
            System.getenv(WRITE_ARTIFACTS_ENV)?.trim()?.equals("true", ignoreCase = true) ?: false

}

